### PR TITLE
Courses

### PR DESCRIPTION
**Pull request by jcypret**
This PR introduces course management for admins and a read-only view for instructors. Also restructures the API to use middleware for authentication and authorization.

### CODE REVIEW

File: seeders/courses.seeder.js [hunk 1]
```diff
@@ -12,13 +12,13 @@ const data = async () => {
   return [
     {
       title: "Programming1",
```
Comment (mdjohns) [inline seeders/courses.seeder.js:14]: This is probably more relevant on the open database PR, but we should probably do another pass over our seed data to make sure things are formatted correctly. I can't think of a reason why we'd leave out the space here?

File: seeders/studentWorkProjects.seeder.js [hunk 2]
```diff
@@ -12,27 +12,27 @@ const data = async () => {
   return [
     {
       name: "Prog1 Assignment 1",
```
Comment (mdjohns) [inline seeders/studentWorkProjects.seeder.js:14]: Same as above, we should circle back on the other PR to make sure we're spelling things out fully.

File: src/components/CourseFormModal.jsx [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            )}
+          </Form.Group>
+          <Form.Group>
+            <Form.Label>Title</Form.Label>
+            <Form.Control
+              type="text"
+              name="title"
+              defaultValue={course?.title}
+              isInvalid={errors.title}
+              ref={register({ required: "Course Title is required" })}
```
Comment (mdjohns) [inline src/components/CourseFormModal.jsx:99]: I'm not sure how deep we want to go on the field validation stuff, but I'm worried the client will hit one of the silly edge cases in the demo. 

I hit two scenarios with the Course Title field that we may or may not care about:

- Entering only space(s) (generic error response from the API, do we want to add any specific error handling for that field?)
- Entering only special characters (valid, able to create a course)

**Review by mdjohns — CHANGES_REQUESTED** [review]
I found two field validation scenarios that we should decide if we want to address. But this looks great!

File: seeders/courses.seeder.js (hunk 1 above)
Comment (I-am-Scott-White) [reply seeders/courses.seeder.js:14]: This was already addressed in a previous commit.

File: seeders/studentWorkProjects.seeder.js (hunk 2 above)
Comment (I-am-Scott-White) [reply seeders/studentWorkProjects.seeder.js:14]: This will be updated in the next push.

File: src/components/CourseFormModal.jsx (hunk 3 above)
Comment (jcypret) [reply src/components/CourseFormModal.jsx:99]: Yeah, that empty field validation is annoying. Really feels like `required: true` should disallow an empty string. That said, sounds like we can avoid both cases you mentioned with a regex matcher, ensuring a non-empty alphanumeric string.

**Comment by jcypret** [comment]
Found a good solution for the validation. Anywhere we care about a guarding against an empty string in the client-side validations, we can add a validate method like this:

```js
ref={register({
  required: "Course Title is required",
  validate: (value) => {
    return !!value.trim() || "Course Title cannot be blank";
  },
  pattern: {
    value: /^[a-zA-Z0-9 ]+$/,
    message: "Course Title must contain only letters and numbers",
  },
})}
```

**Review by mdjohns — COMMENTED** [review]
The field validation looks great!

I'm not sure if it was just me, but I was getting this error on first load of the Courses page:

```
Schema hasn't been registered for model "StudentOutcome".
Use mongoose.model(name, schema)
```
I went ahead and tried it with a production build and saw the same issue. If I navigate to the SOs page first, the Courses load just fine.

I believe this is the first page that uses `ref`. Do you think this is a fluke on my end, or do we need to import `StudentOutcome` in `Course.js` ?

**Comment by jcypret** [comment]
Ah, yeah, this looks like what we talked about on the call last week. Did some more digging today and looks like we need to explicitly import any models we use for `populate`. Grepping the codebase it looks like this is the only instance so far. The nice thing is we can [import without actually declaring a variable](https://github.com/ualr-htcpcp/dcia/pull/62/files#diff-3a1829d4228645d5088d966ee1ab9a3e5ab1d52adf23ddbc5f7926205a708529R4) to load the schema into Mongoose. Pushing changes now ✌️

**Review by mdjohns — APPROVED** [review]

**Pull request merged**